### PR TITLE
Improve performance of initial map list display

### PR DESCRIPTION
**Pull request by moreginger**
I was surprised it took "so long" to show the map list so poked around trying to understand why.

The worst part by far is all the recursive copying that takes place during object init... I was able to make some very small improvements here.

Overall this sped up the time to initial show from 0.72s -> 0.45s. Many of the changes should be of general (small) benefit.

I'll add more comments inline or see the individual commits for details.

### CODE REVIEW

File: LuaMenu/widgets/gui_maplist_panel.lua [hunk 1]
```diff
@@ -176,7 +155,14 @@ local function InitializeControls()
...
 
-	local Configuration = WG.Chobby.Configuration
+	wgChobbyConfig = WG.Chobby.Configuration
+
+	-- Sharing font saves about 20%, but getting the config for the TextBox font isn't straightforward.
+	local fontSpec = wgChobbyConfig:GetFont(2)
+	local sacrificialTextBox = TextBox:New({})
+	fontSpec.font = sacrificialTextBox.font.font
+	sacrificialTextBox:Dispose()
+	listFont = Font:New(fontSpec)
```
Comment (moreginger) [inline LuaMenu/widgets/gui_maplist_panel.lua:165]: Ugh. I couldn't work out a better way to get the text box font, but this was the biggest single performance boost I found.

File: libs/chiliui/chili/controls/control.lua [hunk 2]
```diff
@@ -268,79 +268,42 @@ end
 
 
 function Control:DetectRelativeBounds()
-	--// we need min 2 x-dim coords to define a rect!
-	local numconstraints = 0
-	if (self.x) then
-		numconstraints = numconstraints + 1
+	--// we need min 2 x/y specifiers to define a rect!
+	local x, right, width, y, bottom, height = self.x, self.right, self.width, self.y, self.bottom, self.height
```
Comment (moreginger) [inline libs/chiliui/chili/controls/control.lua:272]: Marginally faster, and clearer?

File: libs/chiliui/chili/controls/editbox.lua [hunk 3]
```diff
@@ -81,21 +91,25 @@ local inherited = this.inherited
...
-	obj._interactedTime = Spring.GetTimer()
+	obj._interactedTime = spGetTimer()
 		--// create font
-	obj.hintFont = Font:New(obj.hintFont)
-	obj.hintFont:SetParent(obj)
+	if obj.hint then
+		obj.hintFont = Font:New(obj.hintFont)
+		obj.hintFont:SetParent(obj)
+	else
+	  obj.hintFont = false
```
Comment (moreginger) [inline libs/chiliui/chili/controls/editbox.lua:100]: Avoiding "getting" the hint font when you have no hint was helpful.

File: libs/chiliui/chili/headers/util.lua [hunk 4]
```diff
@@ -64,26 +64,10 @@ end
 
 function IsRelativeCoord(code)
 	local num = tonumber(code)
-
-	if (type(code) == "string") then
-		return true
-	elseif (num) and ((1/num) < 0) then
-		return true
-	else
+	if num and ((1/num) < 0) then
```
Comment (moreginger) [inline libs/chiliui/chili/headers/util.lua:67]: I reasoned we're more likely to give an pixel constraint than a string constraint? In the map list, it helps.

File: libs/chiliui/chili/headers/util.lua [hunk 5]
```diff
@@ -64,26 +64,10 @@ end
...
 		return false
 	end
-end
-
-function IsRelativeCoordType(code)
```
Comment (moreginger) [inline libs/chiliui/chili/headers/util.lua:77]: This was unused. Maybe I should have kept it, but didn't want to rewrite it blind to match the above.

File: libs/chiliui/chili/headers/util.lua [hunk 6]
```diff
@@ -336,6 +319,18 @@ function table:shallowcopy()
 	return newTable
 end
 
+function table:deepcopy()
```
Comment (moreginger) [inline libs/chiliui/chili/headers/util.lua:322]: deepcopy is faster than merge when there is nothing left to merge, as it avoids lookups on the target table.

File: LuaMenu/widgets/gui_maplist_panel.lua (hunk 1 above)
Comment (GoogleFrog) [reply LuaMenu/widgets/gui_maplist_panel.lua:165]: I'm pretty sure the fonts are cached somewhere in chili.

Comment (moreginger) [reply LuaMenu/widgets/gui_maplist_panel.lua:165]: They are, in fonthandler.

Without objectOverrideFont: 0.5374794 cycles
With objectOverrideFont: 0.45657158 cycles

I'll do some more digging to understand the difference when I have time. I suspect I'll find it's due to object init, again.

**Comment by GoogleFrog** [comment]
If you're improving things in the depths of chili then applying the same improvements ingame would be great.

The game has two versions of chili. The one called `chili` is an exact copy of chobby chili, it is kept synced so porting fixes to it is a non-issue. The one called `chili_old` is a 'fork' of chili from maybe a decade ago that has better performance, at least ingame. It is to do with how things are drawn rather than the core of chili, which I think is mostly the same. I cross-merged the bug fixes, features and API of the two chilis about a year ago in a decent into merge-hell.

https://github.com/ZeroK-RTS/Zero-K/tree/master/LuaUI/Widgets

File: LuaMenu/widgets/gui_maplist_panel.lua (hunk 1 above)
Comment (moreginger) [reply LuaMenu/widgets/gui_maplist_panel.lua:165]: Yeah it looks like the time is in Font:New and calling SetParent on hundreds of Font references.

The way chili styles are applied makes it impossible to get the text box font style another way (without just copy pasta). I'll back this out then. I'll consider messing around with a virtualized list in future work.

**Comment by moreginger** [comment]
> That said, the trick to chili is to reuse controls rather than create new ones. The ingame chili is better in this regard. The creation cost really hurts for the map list and to be truly performant probably requires a system where it doesn't try to load all the data for all the maps at once, perhaps using empty controls for unseen list icons and loading the contents in the background.

Surprisingly, it's not the real content that is the problem AFAICT. It's just the hundreds of control/objects and merge/copy of hundreds of properties (some of them deep merge/copy of tables) that stacks up. I think a virtualized list is probably the right answer, for optimal performance. It might be fun to try and write one, but I'd rather get this merged first if possible.

> If you're improving things in the depths of chili then applying the same improvements ingame would be great.

Happy to try and do that, although yikes at two more versions of chili :) Is this something you'd like to see before accepting this PR?

**Comment by GoogleFrog** [comment]
> Happy to try and do that, although yikes at two more versions of chili :) Is this something you'd like to see before accepting this PR?

No. The scope of this PR is fine. There is also only one additional version.

File: libs/chiliui/chili/controls/editbox.lua (hunk 3 above)
Comment (GoogleFrog) [reply libs/chiliui/chili/controls/editbox.lua:100]: What if the hint is later set?

**Comment by GoogleFrog** [comment]
There was a problem https://github.com/ZeroK-RTS/Chobby/pull/798

**Comment by GoogleFrog** [comment]
Sorry about that @moreginger, I was making a release and this seemed to work in most areas. I'd still like the PR but I don't have the time to do the testing and fixes myself. Perhaps it would be easier to handle as a PR about the map list and a PR about chili internals.

**Pull request merged**